feat: add ES module support

angeloashmore opened the pull request:
The following changes are done in preparation for #10.

- Adds Rollup as a bundler to output CommonJS and ES module. ES module supports tree shaking.
- ~~Updates Typescript and adds `tslib`~~.
- ~~Adds `@ts-ignore` in `pickTrueObjectKeys` ([jump to line](https://github.com/unsplash/ts-imgix/compare/master...angeloashmore:rollup?expand=1#diff-f41e9d04a45c83f3b6f6e630f10117feR80)). Not sure how to fix the type warning. The code is admittedly too complex for me to understand.~~

angeloashmore committed:
feat: add ES module target

OliverJAsh commented on src/index.ts:
```diff
@@ -76,6 +76,7 @@ export type ImgixUrlQueryParams = {
 const pickTrueInObject = <K extends string>(obj: Record<K, boolean>): Partial<Record<K, true>> =>
     pickBy(obj, (_key, value): value is true => value);
 const pickTrueObjectKeys = pipe(
+    // @ts-ignore
```
> * Adds `@ts-ignore` in `pickTrueObjectKeys` ([jump to line](https://github.com/unsplash/ts-imgix/compare/master...angeloashmore:rollup?expand=1#diff-f41e9d04a45c83f3b6f6e630f10117feR80)). Not sure how to fix the type warning. The code is admittedly too complex for me to understand.

How come this wasn't a problem before? Does this PR change the TS version or something? If so it would be good if we could stick to the same TS version and leave that upgrade for another PR.

angeloashmore replied:
Correct, I believe upgrading Typescript caused this compiler error. It has since been restored to the previous version.

OliverJAsh commented on rollup.config.js:
```diff
@@ -0,0 +1,20 @@
+import typescript from 'rollup-plugin-typescript2';
+import autoExternal from 'rollup-plugin-auto-external';
+
+import pkg from './package.json';
```
Nice!

OliverJAsh commented on rollup.config.js:
```diff
@@ -0,0 +1,20 @@
+import typescript from 'rollup-plugin-typescript2';
+import autoExternal from 'rollup-plugin-auto-external';
```
What does this do in our case? I'm new to Rollup

angeloashmore replied:
- `typescript` runs the code through `tsc`
- `autoExternal` sets `dependencies` and `peerDependencies` as external packages so the bundled code leaves them as `import`s. Non-external packages will be bundled in the compiled output.

`rollup-plugin-auto-external` could be replaced with an inlined version. It effectively does the following:

```js
import pkg from './package.json'

// Returns a function that returns true if the provided package name is in `externalArr`.
const makeExternalPredicate = externalArr => {
  if (externalArr.length === 0) {
    return () => false
  }
  const pattern = new RegExp(`^(${externalArr.join('|')})($|/)`)
  return id => pattern.test(id)
}

export default {
  input: 'src/index.ts',
  output: [
    { file: pkg.main, format: 'cjs' },
    { file: pkg.module, format: 'es' },
  ],
  external: makeExternalPredicate([
    ...Object.keys(pkg.dependencies || {}),
    ...Object.keys(pkg.peerDependencies || {}),
  ]),
  plugins: [typescript()],
}
```

OliverJAsh commented on package.json:
```diff
@@ -22,10 +24,14 @@
+    "rollup-plugin-auto-external": "^2.0.0",
+    "rollup-plugin-typescript2": "^0.27.0",
     "source-map-support": "^0.5.9",
+    "tslib": "^1.11.1",
```
How come we need to list this as a dependency now?

angeloashmore replied:
Thanks for pointing this out. It is an unnecessary dependency as it is an implicit dependency of the rollup plugin.

angeloashmore committed:
fix: downgrade typescript and remove tslib

These upgrades/additions were remnants from using @rollup/typescript
before changing to rollup-plugin-typescript2. rollup-plugin-typescript2
does not require tslib.

angeloashmore committed:
fix: use node module resolution

When changing the module compiler option to esnext from commonjs, one
must also define the correct module resolution strategy.

angeloashmore commented:
Trimmed down the PR to just include the Rollup changes.

OliverJAsh commented:
https://github.com/unsplash/ts-imgix/releases/tag/v0.0.12
